### ISSUE TITLE
HDF5 - Mac OS

### ISSUE THREAD

**Issue by alexandrub53** [TRIQS/triqs#393]
I'm trying to install Triqs on my MacOS X El Capitan for the first time. I ran through the installation process as described via homebrew on the website. I've noticed that my two versions of HDF5 are not compatible so I tried to see if I take out the compatibility check from the setup if the code can compile and work. It did compile but I got the following error during the test: 

The following tests FAILED:
	167 - operator_python2 (Failed)
	169 - gf_init (Failed)
	170 - gf_base_op (Failed)
	173 - dos (Failed)
	174 - pade (Failed)
	179 - block_matrix (Failed)
	180 - gf_tensor_valued (Failed)
	181 - histograms (Failed)
	182 - multivar (Failed)
	183 - g_k_om (Failed)
	184 - block2 (Failed)
	185 - wrap_a (Failed)
	188 - wrap_multivar (Failed)
Errors while running CTest
make: *** [test] Error 8

I then tried to change hid_t to int and then to int64_t in this file (4  triqs/python_tools/converters/h5.hpp ) but the same issues appeared. Is there any solution to this at the moment?

**Comment by HugoStrand** [comment]
Dear Alexandru,

The `hid_t` data type should be the correct thing to do if you have a homogeneous HDF5 setup. Could you please provide a bit more info on the breaking tests so I can help you debug? 

The interesting information is stored in your `cmake` build directory under `./Testing/Temporary/LastTest.log` which will show the standard output from the tests.

Even more info you can find in the test subdirectory e.g. `./test/pytriqs/gf_multivar/g_k_om.err` which will contain the full stderr output from the test.

The tests that are failing are seemingly all using hdf5 and python together which means that the probable cause of trouble is that the python module `h5py` on your system is not using the same `hdf5` library as your `triqs` build.

Could you please check the `hdf5` library version used by `h5py` by running:
```python
import h5py
print 'h5py.version.version =', h5py.version.version
print 'h5py.version.hdf5_version =', h5py.version.hdf5_version
print 'h5py.version.api_version =', h5py.version.api_version
print h5py.version.info
```

To check exactly which `hdf5` library is used for `h5py` and `triqs` on Mac OS X you can use `otool -L <executable>`. For `triqs` you can look at any library in the `pytriqs` directory under your `cmake` buld directory, e.g. `otool -L ./pytriqs/gf/gf_fnt.so`. For `h5py` you can run this little python script:

```python
import os
import h5py
# get the path to the h5py module
path_h5py = os.path.abspath(h5py.__file__).strip('__init__.pyc')
# use otool to look at the libraries linked to one of the .so files in h5py, e.g. h5.so
os.system('otool -L ' + path_h5py + 'h5.so')
```
Please check these things and let me know the result.

Best regards, Hugo

**Comment by alexandrub53** [comment]
Dear Hugo,

Thank you for your help!

Yes, so if I run: cmake -DCMAKE_INSTALL_PREFIX=path_to_install_directory ../triqs.src 
without modifying CMakeLists.txt beforehand I would normally get the error message:
  The HDF5 version of the h5py (1.8.18) is different from the hdf5 library
  found for C++ (1.10.0).  You MUST have the same version.
Since I haven't been able to find a way to fix the libraries to be the same using homebrew I edited CMakeLists.txt to not do this check to see if I can make it work. 

 The errors I get are after compiling with two different HDF5. 

If I run your first script I get:

>>> import h5py
>>> print 'h5py.version.version =', h5py.version.version
h5py.version.version = 2.7.0
>>> print 'h5py.version.hdf5_version =', h5py.version.hdf5_version
h5py.version.hdf5_version = 1.8.18
>>> print 'h5py.version.api_version =', h5py.version.api_version
h5py.version.api_version = 1.8
>>> print h5py.version.info
Summary of the h5py configuration
---------------------------------

h5py    2.7.0
HDF5    1.8.18
Python  2.7.13 (default, Apr  4 2017, 08:46:44) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
sys.platform    darwin
sys.maxsize     9223372036854775807
numpy   1.11.0



After running the second:


>>> import os
>>> import h5py
>>> path_h5py = os.path.abspath(h5py.__file__).strip('__init__.pyc')
>>> os.system('otool -L ' + path_h5py + 'h5.so')
/usr/local/lib/python2.7/site-packages/h5py/h5.so:
	@loader_path/.dylibs/libhdf5.10.dylib (compatibility version 13.0.0, current version 13.1.0)
	@loader_path/.dylibs/libhdf5_hl.10.dylib (compatibility version 12.0.0, current version 12.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
0

**Comment by alexandrub53** [comment]
I've also attached files with errors after compiling the code and commenting out the check for HDF5 compatibility CMakeLists.txt and trying to run the code.



[LastTest.log.zip](https://github.com/TRIQS/triqs/files/921127/LastTest.log.zip)
[g_k_om.err.zip](https://github.com/TRIQS/triqs/files/921128/g_k_om.err.zip)

**Comment by HugoStrand** [comment]
Dear Alexandru,

I'm afraid that removing the warning in the `CMakeLists.txt` is not going to remove your problem. The warning is there for a very good reason, simply that mixing `hdf5` libraries in `triqs` do not work, especially the `1.8` and `1.10` versions where the internal types like `hid_t` of `hdf5` has been modified.

I am a bit worried over your output from `h5py` and the library it resolves `@loader_path/.dylibs/libhdf5.10.dylib` is not a hard path so it does not even resolve what library will be used.

The way forward here is really forcing the c++ code in triqs to use the same hdf library as your `h5py` installation. To do this you need the path to the `hdf5` version 1.8 library, try excecuting 

```locate libdf5```

to see the path, and then throw an extra flag to `cmake` to tell it to look for hdf5 in that directory,

``` cmake ... -DHDF5_ROOT=<the path to libhdf5 with /lib/ removed> ```

for specifics on cmake's hdf5 finder script please see:
https://cmake.org/cmake/help/v3.8/module/FindHDF5.html

So in short please re-enable the cmake warning and make it go away by forcing cmake to use the "right" hdf5 library by adding flags to cmake.

Please let me know if this works out.

**Comment by alexandrub53** [comment]
Dear Hugo,

I deleted the version of HDF5 installed by homebrew as there is no way to downgrade it that I could figure out. I then installed HDF5 for UNIX from here, and all the test passed (I used the cmake version): https://support.hdfgroup.org/HDF5/release/cmakebuild518.html 

I cleaned my build folder for triqs, brought the configure file back to normal in src and ran this:
cmake -DHDF5_ROOT=/usr/local/lib/HDF_Group/HDF5/1.8.18  -DCMAKE_INSTALL_PREFIX=.  ../triqs.src 

When the script ran, for the HDF5 detection part I got:
-- -------- HDF5 detection -------------
--  HDF5_LIBRARIES = /usr/local/lib/HDF_Group/HDF5/1.8.18/lib/libhdf5.a;/usr/local/lib/HDF_Group/HDF5/1.8.18/lib/libhdf5_hl.a
-- hdf5 version: 1.8.18


so it seemed ok and the compilation ran. but the make test then gave me this result:
88% tests passed, 23 tests failed out of 189

Total Test time (real) =  22.66 sec

The following tests FAILED:
	 26 - h5_stack (OTHER_FAULT)
	166 - operator_python_test (Failed)
	167 - operator_python2 (Failed)
	169 - gf_init (Failed)
	170 - gf_base_op (Failed)
	171 - gf_fourier (Failed)
	172 - gf_slicing (Failed)
	173 - dos (Failed)
	174 - pade (Failed)
	175 - gf_inplace_112 (Failed)
	176 - gf_transpose (Failed)
	177 - gf_eval (Failed)
	179 - block_matrix (Failed)
	180 - gf_tensor_valued (Failed)
	181 - histograms (Failed)
	182 - multivar (Failed)
	183 - g_k_om (Failed)
	184 - block2 (Failed)
	185 - wrap_a (Failed)
	186 - wrap_B (Failed)
	187 - wrap_g (Failed)
	188 - wrap_multivar (Failed)
	189 - blockgf (Failed)
Errors while running CTest
make: *** [test] Error 8


The result for the first script you sent me is:

5py.version.version = 2.7.0
h5py.version.hdf5_version = 1.8.18
h5py.version.api_version = 1.8
Summary of the h5py configuration
---------------------------------

h5py    2.7.0
HDF5    1.8.18
Python  2.7.13 (default, Apr  4 2017, 08:46:44) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
sys.platform    darwin
sys.maxsize     9223372036854775807
numpy   1.11.0



For the second script:
/usr/local/lib/python2.7/site-packages/h5py/h5.so:
	@loader_path/.dylibs/libhdf5.10.dylib (compatibility version 13.0.0, current version 13.1.0)
	@loader_path/.dylibs/libhdf5_hl.10.dylib (compatibility version 12.0.0, current version 12.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)

I've also attached the two files you mentioned before.

[LastTest.log.zip](https://github.com/TRIQS/triqs/files/924131/LastTest.log.zip)
[g_k_om.err.zip](https://github.com/TRIQS/triqs/files/924132/g_k_om.err.zip)


Thanks again for all your help,
Alex

**Comment by HugoStrand** [comment]
Dear Alexandru,

I am sorry for being unclear.

Let me try to explain the origin of the `hdf5` problems. Currently `triqs` uses two different ways of dealing with `hdf5` data files. On the c++ side it uses the `hdf5` library while on the python side it uses the `h5py` python module (that in turn uses the `hdf5` library).

Now, it is central that the `hdf5` library that is used in the two cases, 1. within the c++ code and 2. in python through `h5py` is _exactly_ the same.

Thus, it does not help that you install a separate `hdf5` library and link it with `triqs`, unless you also at the same time install the python module `h5py` using the same `hdf5` library. In your current setup you statically link `triqs` to one `hdf5` library (`/usr/local/lib/HDF_Group/HDF5/1.8.18/lib/libhdf5.a) while your `h5py` module dynamically links to `@loader_path/.dylibs/libhdf5.10.dylib`.

The errors you now are getting (missing symbol `_SZ_encoder_enabled` see `g_k_om.err`) seems to be related to the hdf5 library you installed by hand, and that it needs some additional libraries (libsz.a and libz.a) since it was compiled with compression library support.

I would recommend you to try to get this working using either `brew` or `MacPorts` (which is what I'm using).

**Comment by alexandrub53** [comment]
After a bit of back and forth (not shown here), the solution to this problem is pretty simple.

When running the pip install command, pip actually downloads a binary and doesn't compile h5py using hdf5 installed on the system. So the h5py binary that is downloaded is compiled with an older version of hdf5 than is installed by brew. This is also the case for mpi4py. The way to solve this is to force pip to download the required files and compile h5py and mpi4py using the local brew configuration.


so I uninstalled the h5py and mpi4py installed by pip with pip uninstall h5py etc then ran

pip install --no-binary=h5py h5py
pip install  --no-binary=mpi4py mpi4py


Following this, the configuration and tests all worked.
